packages/flutter/lib/src/material/notch_computers.dart

```diff
@@ -0,0 +1,158 @@
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+/// Computes notches in the outline of a shape.
+///
+/// Typically used to compute notches by a [BottomAppBar].
+abstract class NotchComputer {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const NotchComputer();
+
+  /// Returns a path for a notch in the outline of a shape.
+  ///
+  /// The path makes a notch in the host shape that can contain the guest shape.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  ///
+  /// The returned path may pass anywhere, including inside the guest bounds area,
+  /// and may contain multiple subpaths. The returned path ends at `end` and does
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
+  Path compute(Rect host, Rect guest, Offset start, Offset end);
+}
+
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
+/// horizontal edge with a circle circumference.
+///
+/// See also: [compute].
+class CircularNotchComputer implements NotchComputer {
+  /// Constructs a [CircularNotchComputer] that can be used for multiple notch
+  /// computations.
+  const CircularNotchComputer({
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
```

Comment:
It might be clearer to just say that we're expanding the guest's implicit radius by notchMargin. You'd need to explain how we come up with the guest's implicit radius of course.